pkg/controllers/workapplier/backoff.go

```diff
@@ -0,0 +1,222 @@
+	}
+
+	if maxDelaySeconds < minFastDelaySeconds {
+		maxDelaySeconds = minFastDelaySeconds
+		klog.V(2).InfoS("maxDelaySeconds is below the minimum value (%d seconds); set it to the minimum value instead", minFastDelaySeconds)
+	}
+	if maxDelaySeconds > maxMaxDelaySeconds {
+		maxDelaySeconds = maxMaxDelaySeconds
+		klog.V(2).InfoS("maxDelaySeconds is above the maximum value (%d seconds); set it to the maximum value instead", maxMaxDelaySeconds)
+	}
+
+	return &RequeueFastSlowWithExponentialBackoffRateLimiter{
+		requeueCounter:                  make(map[types.NamespacedName]int),
+		lastTrackedGeneration:           make(map[types.NamespacedName]int64),
+		lastTrackedProcessingResultHash: make(map[types.NamespacedName]string),
+		maxFastAttempts:                 maxFastAttempts,
+		fastDelay:                       time.Duration(fastDelaySeconds * float64(time.Second)),
+		exponentialBase:                 exponentialBase,
+		maxDelay:                        time.Duration(maxDelaySeconds * float64(time.Second)),
+	}
+}
+
+// When returns the duration to wait before requeuing the item.
+//
+// The implementation borrows from the workqueue package's various rate limiter implementations.
+func (r *RequeueFastSlowWithExponentialBackoffRateLimiter) When(work *fleetv1beta1.Work, bundles []*manifestProcessingBundle) time.Duration {
+	r.mu.Lock()
+	defer r.mu.Unlock()
+
+	namespacedName := types.NamespacedName{
+		Namespace: work.Namespace,
+		Name:      work.Name,
+	}
+
+	// Check if the Work object has been tracked before; if so, verify if its has a generation change
+	// or the processing result has changed.
+	lastTrackedGen, isTracked := r.lastTrackedGeneration[namespacedName]
+	lastTrackedProcessingResHash := r.lastTrackedProcessingResultHash[namespacedName]
+	curProcessingResHash, hashErr := computeProcessingResultHash(work, bundles)
+	switch {
+	case !isTracked:
+		// The Work object has never been tracked before. No action is needed.
+	case lastTrackedGen != work.Generation:
+		// A new generation of the Work object has been observed.
+		// Reset the requeue counter.
+		r.requeueCounter[namespacedName] = 0
+	case hashErr != nil:
+		// Cannot compute the hash of the processing result. Normally this should not occur.
+		// Reset the requeue counter just in case.
+		r.requeueCounter[namespacedName] = 0
+	case lastTrackedProcessingResHash != curProcessingResHash:
+		// The processing result has changed.
+		// Reset the requeue counter.
+		r.requeueCounter[namespacedName] = 0
+	default:
+		// The Work object has been tracked before and its generation and processing result have not changed.
+		// No action is needed.
+	}
+
+	// Update the last tracked generation and processing result for the Work object.
+	r.lastTrackedGeneration[namespacedName] = work.Generation
+	r.lastTrackedProcessingResultHash[namespacedName] = curProcessingResHash
```

Comment:
what if hashErr != nil?

Reply:
But if you prefer it to be more explicit I could make the edit.